src/ring_theory/graded_algebra/basic.lean

```diff
@@ -77,4 +80,101 @@ def graded_algebra.decompose : A ≃ₐ[R] ⨁ i, 𝒜 i := alg_equiv.symm
   (graded_algebra.decompose 𝒜).symm (direct_sum.of _ i x) = x :=
 direct_sum.submodule_coe_alg_hom_of 𝒜 _ _
 
+/-- The projection maps of graded algebra-/
+def graded_algebra.proj (𝒜 : ι → submodule R A) [graded_algebra 𝒜] (i : ι) : A →ₗ[R] A :=
+(𝒜 i).subtype.comp $
+  (dfinsupp.lapply i).comp $
+  (graded_algebra.decompose 𝒜).to_alg_hom.to_linear_map
+
+@[simp] lemma graded_algebra.proj_apply (i : ι) (r : A) :
+  graded_algebra.proj 𝒜 i r = (graded_algebra.decompose 𝒜 r : ⨁ i, 𝒜 i) i := rfl
+
+lemma graded_algebra.proj_mem (i : ι) (r : A) :
+  graded_algebra.proj 𝒜 i r ∈ 𝒜 i := (graded_algebra.decompose 𝒜 r i).2
+
+/-- The support of `r` is the `finset` where `proj R A i r ≠ 0 ↔ i ∈ r.support`-/
+def graded_algebra.support [Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0)]
+  (r : A) : finset ι :=
+(graded_algebra.decompose 𝒜 r).support
+
+lemma graded_algebra.proj_recompose (a : ⨁ i, 𝒜 i) (i : ι) :
+  graded_algebra.proj 𝒜 i ((graded_algebra.decompose 𝒜).symm a) =
+  (graded_algebra.decompose 𝒜).symm (direct_sum.of _ i (a i)) :=
+by rw [graded_algebra.proj_apply, graded_algebra.decompose_symm_of, alg_equiv.apply_symm_apply]
+
+variable [Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0)]
+
+lemma graded_algebra.mem_support_iff
+  (r : A) (i : ι) :
+i ∈ graded_algebra.support 𝒜 r ↔ (graded_algebra.proj 𝒜 i r ≠ 0) :=
+begin
+  rw [graded_algebra.support, dfinsupp.mem_support_iff, graded_algebra.proj_apply],
+  simp only [ne.def, submodule.coe_eq_zero],
+end
+
+lemma graded_algebra.as_sum (r : A) :
+  r = ∑ i in graded_algebra.support 𝒜 r, graded_algebra.proj 𝒜 i r :=
+begin
+  conv_lhs { rw [←@graded_algebra.right_inv ι R A _ _ _ _ _ 𝒜 _ r,
+    graded_algebra.decompose'_def,
+    ←direct_sum.sum_support_of _ (graded_algebra.decompose 𝒜 r), linear_map.map_sum] },
+  unfold graded_algebra.support,
+  apply finset.sum_congr rfl,
+  intros i hi,
+  rw [direct_sum.submodule_coe_of],
+  refl,
+end
+
+lemma graded_algebra.mul_proj (r r' : A) (i : ι) :
+  graded_algebra.proj 𝒜 i (r * r') =
+  ∑ ij in finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i)
+    ((graded_algebra.support 𝒜 r).product (graded_algebra.support 𝒜 r')),
+    (graded_algebra.proj 𝒜 ij.1 r) * (graded_algebra.proj 𝒜 ij.2 r') :=
```

Comment:
Should this be rewrite in terms of `decompose` as well

Reply:
yes, I think they all should. If you have two different ways to spell the same thing (`decompose r i` vs `proj i r`), you should pick one as the normal for, and state all the lemmas in terms of that one.